core/src/main/java/ai/timefold/solver/core/impl/score/stream/common/ForEachFilteringCriteria.java

```diff
@@ -1,28 +1,33 @@
 package ai.timefold.solver.core.impl.score.stream.common;
 
-import java.util.function.Function;
+import java.util.function.BiFunction;
 import java.util.function.Predicate;
 
 import ai.timefold.solver.core.impl.domain.entity.descriptor.EntityDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.declarative.ConsistencyTracker;
 
 import org.jspecify.annotations.NullMarked;
 import org.jspecify.annotations.Nullable;
 
 @NullMarked
 public enum ForEachFilteringCriteria {
-    ASSIGNED_AND_CONSISTENT(entityDescriptor -> entityDescriptor.getEntityForEachFilter().getAssignedAndConsistentPredicate()),
-    CONSISTENT(entityDescriptor -> entityDescriptor.getEntityForEachFilter().getConsistentPredicate()),
-    ALL(ignored -> null);
+    ASSIGNED_AND_CONSISTENT((consistencyTracker, entityDescriptor) -> entityDescriptor.getEntityForEachFilter()
+            .getAssignedAndConsistentPredicate(consistencyTracker)),
+    CONSISTENT((consistencyTracker, entityDescriptor) -> entityDescriptor.getEntityForEachFilter()
+            .getConsistentPredicate(consistencyTracker)),
+    ALL((ignored1, ignored2) -> null);
 
-    private final Function<EntityDescriptor<?>, @Nullable Predicate<Object>> entityDescriptorToPredicateFunction;
+    private final BiFunction<ConsistencyTracker<?>, EntityDescriptor<?>, @Nullable Predicate<Object>> entityDescriptorToPredicateFunction;
 
-    ForEachFilteringCriteria(Function<EntityDescriptor<?>, Predicate<Object>> entityDescriptorToPredicateFunction) {
+    ForEachFilteringCriteria(
+            BiFunction<ConsistencyTracker<?>, EntityDescriptor<?>, Predicate<Object>> entityDescriptorToPredicateFunction) {
         this.entityDescriptorToPredicateFunction = entityDescriptorToPredicateFunction;
     }
 
     @Nullable
     @SuppressWarnings("unchecked")
-    public <A> Predicate<A> getFilterForEntityDescriptor(EntityDescriptor<?> entityDescriptor) {
-        return (Predicate<A>) entityDescriptorToPredicateFunction.apply(entityDescriptor);
+    public <A> Predicate<A> getFilterForEntityDescriptor(ConsistencyTracker<?> consistencyTracker,
+            EntityDescriptor<?> entityDescriptor) {
+        return (Predicate<A>) entityDescriptorToPredicateFunction.apply(consistencyTracker, entityDescriptor);
```

Comment:
For node-sharing, is this guaranteed to return the same instance every time?

Looking at where this leads, it'll be `entityConsistencyState::isEntityConsistent` - and that will be a different object every time.

For peace of mind and the future generations, there should be a test guaranteeing this behavior.

Reply:
We already have node sharing tests; this predicate is cached inside `ConstraintNodeBuildHelper` since it returns different instances (as it depends on `ConsistencyTracker`). `ForEach` also wraps the predicate inside `ForEachFilteringCriteriaPredicateFunction` that implements equality/hashCode based on entity class + criteria.